# Corgias

<div align="center">
<p align="center">
    <img src="CORGIAS.png?raw=true?" alt="corgias-logo" width="300">
   <br> This logo was generated with the aid from AI (DALLÂ·E).
</p>
<h1>CORGIAS</h1>
<h3>CORrelated Genes Identifier by considering Ancestral State</h3>
</div>

CORGIAS is a phylogenetic profiling tool for a large-scale dataset comprising of thousands and tens of thousands of orthologs and genomes. As co- and anti-correlated orthologs are expexted to be functionally related, CORGIAS can help functional annotation of orthologs, especially those showing no sequence similarity to functionally known genes.

## Installation

Corgias can be installed on Linux system.

```bash
git clone https://github.com/XXXX
cd corgias

# Optional, but recommended to install in the virtual environment
python -m venv .venv
. .venv/bin/activate
python -m pip install .
```

## Usage

CORGIAS provides three subcommand. `profiling` calculate values necessary for evaluating correlation of all ortholog pairs. Two our new profiling methods, Ancestral steate adjustment (ASA) and Simultaneous Evolutionary Test (SEV) requre the ancestral state of orthologs and `asr` prepared them. `stat` performs statistical tests using values calculated by `profiling`.

### Ancestral state construction (ASR)

Subcommand `asr` needs a rooted species tree in newick format and ortholog table in csv or csv-like format (separator can be sepcified with -s/--separator option). This is just a wrapper conductiong `pastml` parallely.

```
$corgias asr -h
usage: corgias asr [-h] -t TREE -d DATA [-i ID_INDEX] [-s SEPARATOR] --work_dir WORK_DIR [-c CORES] [--test TEST] [--tmp TMP] [--keep]

	Prepare trees with ancestral presence/absence states of ortholog for ASA or SEV profiling.
	The ortholog table should be a CSV-like file but each ortholog is assmued to be evolved independently.
	Example usage:
		corgias asr -t tree.nwk -d orthologs.csv -i 0 -s "," -o pastml_result -c 4 --predictio_method ML

	Note: Recostruction should be performed by a maximum-likelihood (DOWNPASS) and maximum-parsimony method (ACCTRAN)
	      for ASA and SEV, respectively.

options:
  -h, --help            show this help message and exit
  -t TREE, --tree TREE
  -d DATA, --data DATA
  -i ID_INDEX, --id_index ID_INDEX
  -s SEPARATOR, --separator SEPARATOR
  --work_dir WORK_DIR
  -c CORES, --cores CORES
  --test TEST
  --tmp TMP
  --keep
```

In addition to the above, `pastml` options can be acceptable (for example, `--upload_to_itol`).

### Phylogenetic profiling
Six phylogenetic profiling mehtods, naive, run length encoding (rle), cladewise adjustment (cwa), ansestral state adjustment (asa), cotransitions (cotr), simultanous evolution test (SEV) are available in subcommand `profiling`. Required input files vary by the methods<br>
**naive**: a ortholog table in csv format. <br>
**rle, cwa, cotr**: a ortholog table in csv format and a rooted species tree in newick format. <br>
**asa, sev**: A resultant folder of `asr` and a rooted species tree in newick format.<br><br>

It is highly recommended to run with `--test` with a small number before running with a large dataset.
```
$corgias profiling -h
usage: corgias profiling [-h] -m {naive,rle,cwa,asa,cotr,sev} [-og OG_TABLE] [-a ASR_FOLDER] -o OUTPUT [-t TREE] [-c CORES] [--ignore_branch] [--gpu]
                         [-nb NUM_BLOCKS] [--test TEST]

	Perform phylogenetic profiling using a ortholog table (naive, rle, cwa, cotr)
	a species tree (rle, cwa, cotr) and/or, ancestral state reconstruction results (asa, sev)
	Example usages: 
		corgias profiling -m naive -og orthologs.csv -o naive_out.csv -c -4 --gpu -nb 4
		corgias profiling -m rle -og orthologs.csv -t tree.nwk -o rle_out.csv -c 4 
		corgias profiling -m cwa -og orthologs.csv -t tree.nwk -o cwa_out.csv -c 4 
		corgias profiling -m asa -a pastml_result_folder -t tree.nwk -o asa_out.csv -c 4 
		corgias profiling -m cotr -og orthologs.csv -t tree.nwk -o cotr_out.csv -c 4 
		corgias profiling -m sev --a pastml_result_folder -t tree.nwk -o sev_out.csv -c 4 

	Note: with --test 5, Run test will start using five orthologs. 

options:
  -h, --help            show this help message and exit
  -m {naive,rle,cwa,asa,cotr,sev}, --method {naive,rle,cwa,asa,cotr,sev}
  -og OG_TABLE, --og_table OG_TABLE
  -a ASR_FOLDER, --asr_folder ASR_FOLDER
  -o OUTPUT, --output OUTPUT
  -t TREE, --tree TREE
  -c CORES, --cores CORES
  --ignore_branch
  --gpu
  -nb NUM_BLOCKS, --num_blocks NUM_BLOCKS
  --test TEST

```



### Statictical test
`stat` only requires the file generated by `profiling` as input.
```
corgias stat [-h] -i INPUT -m {naive,rle,cwa,asa,cotr,sev} [-o OUTPUT] [-d {both,correlation,anti-correlation}] [-c CORES] [-t THRETHOLD]
                    [-s {bonferroni,sidak,holm-sidak,simes-hochberg,hommel,fdr_bh,fdr_by,fdr_tsbh,fdr_tsbky}] [--only_signif]

	Conduct statistical tests for phylogenetic profiling results.
	Example usage:
		corgias stat -i profiling_result.csv -m naive -o stat_out.csv -c 4 

options:
  -h, --help            show this help message and exit
  -i INPUT, --input INPUT
  -m {naive,rle,cwa,asa,cotr,sev}, --method {naive,rle,cwa,asa,cotr,sev}
  -o OUTPUT, --output OUTPUT
  -d {both,correlation,anti-correlation}, --direction {both,correlation,anti-correlation}
  -c CORES, --cores CORES
  -t THRETHOLD, --threthold THRETHOLD
  -s {bonferroni,sidak,holm-sidak,simes-hochberg,hommel,fdr_bh,fdr_by,fdr_tsbh,fdr_tsbky}, --statistical_test {bonferroni,sidak,holm-sidak,simes-hochberg,hommel,fdr_bh,fdr_by,fdr_tsbh,fdr_tsbky}
  --only_signif
```

## License

This project is licensed under the GPL3.0 License. See the `LICENSE` file for details.

## Citation

If you use CORGIAS in your research, please cite it as follows:

```
Yuki Nishimura, Kimiho Omae, Kento Tominnaga, Wataru Iwasaki.
CORGIAS: identifying correlated gene pairs by considering evolutionary history in a large-scale prokaryotic genome dataset
[Journal Name], 2025, [DOI]
```

## Contact

Yuki Nishimura (The University of Tokyo) yuki-nishimura@g.ecc.u-tokyo.ac.jp